DOCS-7946-content added (#243)
mulesoft/docs-api-manager: modules/ROOT/pages/configure-uri-template-regex.adoc

= Configure URI Template Regex
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: policy, custom, ootb, offline, resource level

//When you do {blah blah}, you must configure the URI template regex. -why do we need to confiugure regex?

When configuring the URI template regex, the base path of the Mule runtime engine (Mule) application that implements (or proxies) the API must be considered. This is because Mule 4 resource-level policies are dependent on the location where an API is deployed.

Let’s look at the following example to understand this better:

image:dhsdfksh.png[]

Imagine that this API specification is implemented in the following Mule application:
+
[source,text,linenums]
----

<api-gateway:autodiscovery apiId="1" flowRef="proxy"/>

<http:listener-config name="httpConfig" basePath="api">
   <http:listener-connection host="localhost" port="80"/>
</http:listener-config>

<flow name="proxy">
   <http:listener config-ref="httpConfig" path="v1/*" />
   <logger />
</flow>
----

> In this example, the `HTTP listener-config` element defines a basePath attribute and the HTTP listener element also defines an additional base path, v1, before the trailing ”*” in the path. The concatenation of these two paths is referred to as base path.. The base path in the example now is `/api/v1`.

Reviewer: Extra "." after "referred to as base path".

Author: Done! Thanks!

== Ignoring Base Path

Including the base path in the full path of the resource when applying resource-level policies forces a coupling between the API in design time and the API in runtime. This is because you are required to know the base path where the API is going to reside when applying resource-level policies.

To avoid this coupling issue, a new configuration is introduced in Mule 4.3.0, which  allows the Mule Policy Engine to ignore the base path of the API resource when matching the incoming HTTP request with the URI Template Regex the policy has configured.

The configuration is exposed on the Mule application implementing the API, to which the policies are applied:

`<api-gateway:autodiscovery apiId="1" flowRef="flow_api" ignoreBasePath="true"/>`

When the `ignoreBasePath` attribute is set to `true`, the base path where the API is deployed is not considered anymore. When the Mule application is configured to ignore the base path to protect the first resource of the API specification, the URI template regex should look like this:

image:ljaljdlfs.png[]

> In this case, the Preview Resource Matching option matches correctly with the resource being protected. The resource inturn also matches with the runtime behavior.

Reviewer: inturn?

Author: Done! Thanks!

> = Configure URI Template Regex

Reviewer 2: Having gone further in this topic, I now wonder if it does not need a subtopic inserted:

=Configure a URI Template Regex
  Explains the regex and the concept of base path; how it is used and that sometime it should be ignored.
==Example Defining a Base Path
==When to Ignore the Base Path

Maybe this is not optimal, but it seems odd to have a huge example of all the parts (paths, policies, API resources...) of the URI template regex, but then a solitary subtext about ignoring the base path?

Author: Kept in mind! Done! Thanks!

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: policy, custom, ootb, offline, resource level

> You can specify a regular expression (regex) to include multiple paths when you apply a policy to a subset of resources exposed by an API. The URI Template Regex field in the policy enables you to specify this expression.

Reviewer 2: Is this field name in syntax? If so, please format whatever portion of that name that is verbatim as syntax.

Author: Done! Thanks!

> When configuring the URI template regex, the base path of the Mule runtime engine (Mule) application that implements (or proxies) the API must be considered. This is because Mule 4 resource-level policies are dependent on the location where an API is deployed.

Reviewer 2: Misplaced modifier (When configuring...the base path > When configuring...you)
```suggestion
When configuring the URI template regex, you must consider the base path of the Mule runtime engine (Mule) application that implements (or proxies) the API, because Mule 4 resource-level policies depend on where an API is deployed.
```
Would users that are NOT using Mule 4 be reading this? If so, should we say:
```suggestion
When configuring the URI template regex, if you are using Mule 4, you must consider the base path of the Mule runtime engine (Mule) application that implements (or proxies) the API. This is because Mule 4 resource-level policies are dependent on the location where an API is deployed.
```

Author: Done! Thanks!

> Let’s look at the following example to understand this better:

Reviewer 2: This kind of language might belong in Citizen, but I think we do not want to refer to "us" in our normal docs:
```suggestion
Consider the following example:
```

Author: Done! Thanks!

.This diagram displays the list of methods used to specify the resource subset to apply to a policy.
image:configure-regex-resources-list.png[Policy Resource List]

Imagine that this API specification is implemented in the following Mule application:
+
[source,text,linenums]
----

<api-gateway:autodiscovery apiId="1" flowRef="proxy"/>

<http:listener-config name="httpConfig" basePath="api">
   <http:listener-connection host="localhost" port="80"/>
</http:listener-config>

<flow name="proxy">
   <http:listener config-ref="httpConfig" path="v1/*" />
   <logger />
</flow>
----

> In this example, the `HTTP listener-config` element defines a basePath attribute and the HTTP listener element also defines an additional base path, v1, before the trailing ”*” in the path. The concatenation of these two paths is referred to as base path. The base path in the example now is `/api/v1`.

Reviewer 2: ```suggestion
In this example, `http:listener-config` defines both a `basePath` attribute and an additional path, `"v1/*"`. The concatenation of these two paths is referred to as the _base path_, which is `/api/v1`.
```

Reviewer 2: How did this boldface come about? Is this going to be a problem in the transformed doc, having text between two * *?

Author: Done! Thanks!

> To invoke the first resource of the API, the base path must be used in combination with the resource path `resource-1: http://localhost:80/api/v1/resource-1`.

Reviewer 2: Does this refer to `<api-gateway:autodiscovery apiId="1" flowRef="proxy"/>`?
If there is just one API resource in this example, why refer to it as "the first"?

Reviewer 2: passive voice

Author: Done! Thanks!

> Subsequently, if you want to protect only the first resource (resource-1) with any policy, you must use a resource-level policy. Therefore, the URI template regex that you specify when applying the policy must also include the base path in it:

Reviewer 2: Similarly to referring to the "first," why use "only the first resource" when that is the only one shown?
Does this mean the same thing:
```suggestion
Subsequently, if you want to protect `resource-1` with a policy, you must use a resource-level policy. Therefore, the URI template regex that you specify when applying the policy must also include the base path in it:
```

Author: Done! Thanks!

.This diagram illustrates how to configure a URI template regex.
image:configure-uri-regex-example.png[Configure URI Template Regex Example]

> If  the API is of RAML or OAS type, the `Preview Resource Matching` option does not show any resource as being matched, which differs from the actual runtime behavior.

Reviewer 2: extra space
```suggestion
If the API is of RAML or OAS type, the `Preview Resource Matching` option does not indicate that any resource is being matched, which differs from the actual runtime behavior.
```

Author: Done! Thanks!

> You can also  use a regex to avoid having to explicitly specify the base path, for example, `.*/resource-1`. However, when you have resources with the same name at different levels of the API definition, you might encounter issues. This is because the regex will match with both `/api/v1/resource-1`, and `/api/v1/resources/resource-1`, which makes it an indecisive solution.

Reviewer 2: extra space
```suggestion
You can also use a regex to avoid having to explicitly specify the base path, for example, `.*/resource-1`. However, if you have resources with the same name at different levels of the API definition, the regex will match all of them, resulting in issues.
```

Author: Done! Thanks!

> == Ignoring Base Path

Reviewer 2: ```suggestion
== When to Use ignoreBasePath
```
?

Reviewer 2: COPYEDITED ONLY from this point onward, to meet deadline.
This should probably have an organizational and content review.

Author: Done! Thanks!

> Including the base path in the full path of the resource when applying resource-level policies forces a coupling between the API in design time and the API in runtime. This is because you are required to know the base path where the API is going to reside when applying resource-level policies.

Reviewer 2: "forces a coupling between...design time and...runtime"?
Is this saying that you must know the deployment base path for a production version when you are still in development?
What do you mean by "coupling"? Is this really a "dependency"?
(Oops...copyeding only....)

Author: Not sure. Let's revisit. Done! Thanks!

> To avoid this coupling issue, a new configuration is introduced in Mule 4.3.0, which  allows the Mule Policy Engine to ignore the base path of the API resource when matching the incoming HTTP request with the URI Template Regex the policy has configured.

Reviewer 2: ```suggestion
To avoid this issue, Mule 4.3.0 includes a new configuration that allows the Mule policy engine to ignore the base path of the API resource when matching the incoming HTTP request with the policy's URI template regex.
```

Author: Done! Thanks!

> The configuration is exposed on the Mule application implementing the API, to which the policies are applied:

Reviewer 2: The policies are applied to the API or the Mule app?

Author: Done! Thanks!

`<api-gateway:autodiscovery apiId="1" flowRef="flow_api" ignoreBasePath="true"/>`

> When the `ignoreBasePath` attribute is set to `true`, the base path where the API is deployed is not considered anymore. When the Mule application is configured to ignore the base path to protect the first resource of the API specification, the URI template regex should look like this:

Reviewer 2: ```suggestion
When the `ignoreBasePath` attribute is set to `true`, the base path where the API is deployed is not considered anymore. When the Mule application is configured to ignore the base path to protect the first resource of the API specification, the URI template regex looks like this:
```

Author: Done! Thanks!

> .This diagram illustrates how to specify Preview Resource Matching.

Reviewer 2: ```suggestion
.This diagram illustrates how to specify preview resource matching.
```

Author: Done! Thanks!

image:configure-uri-regex-example.png[Specify Preview Resource Matching]

> In this case, the Preview Resource Matching option matches correctly with the resource being protected. The resource also matches with the runtime behavior.

Reviewer 2: ```suggestion
In this case, the Preview Resource Matching option matches the resource being protected. The resource also matches the runtime behavior.
```

Author: Done! Thanks!

> Prerequisites

Reviewer 2: Shouldn't this be a subhead?

Author: Bolded! Done! Thanks!

To implement this feature, ensure that your Mule application:

* Is running in Mule 4.3.0

> * Contains the Mule HTTP Connector 1.5.15 or above

Reviewer 2: Do you mean Anypoint Connector for HTTP (HTTP Connector)?

And "above" > "later"

Author: Done! Thanks!

> For backwards compatibility reasons, this feature is disabled unless the attribute is explicitly set to `true`. If your Mule applications are running in Mule versions prior to 4.3.0 and have resource-level policies correctly configured, you do not need to make any changes to migrate to Mule 4.3.0 or later.

Reviewer 2: ```suggestion
To achieve backward compatibility, this feature is disabled unless the attribute is explicitly set to `true`. If your Mule applications are running in Mule versions prior to 4.3.0 and have resource-level policies correctly configured, you do not need to make any changes to migrate to Mule 4.3.0 or later.
```

Author: Done! Thanks!

The latest versions of API proxies available in API Manager are preconfigured to ignore the base path. To avoid breaking existing resource-level policies that are implemented using previous versions of API proxies, the following major versions of API proxies are released with this feature:

* HTTP Proxy 2.0.0
* HTTP Proxy 3.0.0

== Feature Limitations

> The root path of an API must be well defined to calculate the base path of the API. To do so, ensure the path attribute of the HTTP listener in the Mule Application ends with a `*`. 

Reviewer 2: ```suggestion
The root path of an API must calculate the base path of the API. Ensure that the path attribute of the HTTP Listener in the Mule application ends with `*`. 
```

Author: Done! Thanks!

> Otherwise, the resource-level policy matching works as though the ignore base path attribute is set to `false`.

Reviewer 2: ```suggestion
Otherwise, resource-level policy matching works as if `ignoreBasePath` is set to `false`.
```

Author: Done! Thanks!
